### ISSUE TITLE
LaTeX equation previews

### ISSUE THREAD

**Issue by kaibagley** [xenodium/chatgpt-shell#274]
Hey,

I'm not an Emacs expert so I'm sure I'm missing an obvious solution, but it would be nice to have latex equation outputs automatically be rendered after the output is complete.

Thanks!

**Comment by xenodium** [comment]
Thanks for filing. Could you please include a sample response? Is it normally inside a code block? Elsewhere?

**Comment by kaibagley** [comment]
Sure:

ChatGPT(chatgpt-4o-latest/General)> What is the PDF for a normal distribution?

> The **Probability Density Function (PDF)** for a normal (Gaussian) distribution is expressed mathematically as:
> 
> \[
> f(x | \mu, \sigma) = \frac{1}{\sqrt{2 \pi \sigma^2}} e^{-\frac{(x - \mu)^2}{2 \sigma^2}}
> \]
>

**Comment by xenodium** [comment]
Thanks. In other Emacs modes, what do you use for rendering LaTeX?

**Comment by kaibagley** [comment]
I use it in org(-roam) for note taking. I run `org-latex-preview`, which calls `latex` and does some dvipng magic I think. It works out of the box with MiKTeX on Windows, and TeXLive on Arch Linux

**Comment by xenodium** [comment]
We can add post-execution actions via `chatgpt-shell-after-command-functions`.

I had a look at `org-latex-preview` to see what I could reuse. Probably not the cleanest but this kinda does the job:

```emacs-lisp
(add-hook 'chatgpt-shell-after-command-functions
          (lambda (command output success)
            (let ((file (buffer-file-name (buffer-base-buffer)))
                  ;; Silence org-element-at-point warnings.
                  (major-mode 'org-mode))
              (save-excursion
                (org-format-latex
                 (concat org-preview-latex-image-directory "org-ltximg")
                 (point-min) (point-max)
                 temporary-file-directory
                 'overlays nil 'forbuffer org-preview-latex-default-process)))))
```

![latex](https://github.com/user-attachments/assets/be1e38ae-93c4-4528-a0d8-5a5356d1e18d)

**Comment by kaibagley** [comment]
Works perfectly for me, thanks so much for your help I really appreciate it. Should I close this issue?

**Comment by kaibagley** [comment]
I just started using the Anthropic API and Claude seems to default to ``` fences for eqauations rather than LaTeX

**Comment by xenodium** [comment]
Could you share entire output please?

**Comment by xenodium** [comment]
Could you share entire output please?

**Comment by xenodium** [comment]
Also, have you tried creating a system prompt which explicitly requests LaTeX? This should hopefully work across providers.

See chatgpt-shell-system-prompts and chatgpt-shell-swap-system-prompt

**Comment by xenodium** [comment]
> Claude(3-5-sonnet-20240620/tl;dr)>

tl;dr system prompt prolly isn’t helping us here wrt to blocks. 

> Claude(3-5-sonnet-20240620/Mathematics)> 

Would you be ok with me adding your new prompt to the default list?

> Works perfectly for me, thanks so much for your help I really appreciate it.

Glad to hear it! Consider supporting this effort to make it sustainable https://github.com/sponsors/xenodium

**Comment by kaibagley** [comment]
1. Very true, wasn't paying attention...
2. No problem at all, I'm sure it could use some refining though. It works much better with ChatGPT than Claude too..
3. Done!

**Comment by xenodium** [comment]
Thank you! ☺️🫡

**Comment by zstg** [comment]
https://github.com/user-attachments/assets/23ca3130-7eb1-4eb4-b3e2-b7c5274c05a6

How do I know _what_ to tweak the system prompt so as to render using LaTeX? I've tried it with all the default system prompts to no avail.

**Comment by xenodium** [comment]
Did you see @kaibagley's snippet in previous comment? https://github.com/xenodium/chatgpt-shell/issues/274#issuecomment-2547606253

```emacs-lisp
  (add-to-list 'chatgpt-shell-system-prompts
               (cons "Mathematics"
                     "The user is a mathematician with very limited time.
You treat their time as precious. You do not repeat obvious things, including their query.
You are as concise as possible in responses.
You never apologize for confusions because it would waste their time.
All mathematical outputs must be in proper LaTeX format.
Use \\( and \\( for in-line equations, and \\[ and \\] for equation environments.
All mathematical delimiters must be on their own line.
All mathematical outputs are formally expressed in formal mathematical notation.
Don't give approximations or pronounciations for constants or variables.
You give formal definitions for all variables.
You keep your answers succinct and to-the-point.")
```

You then have to invoke `chatgpt-shell-swap-system-prompt` and choose `Mathematics`.

**Comment by zstg** [comment]
Apologies. I got it working. I thought I had deleted the above comment. Sorry.

**Comment by xenodium** [comment]
@kaibagley @zstg 

I've added the Mathematics system prompt and enabled latex rendering by default in https://github.com/xenodium/chatgpt-shell/commit/420388ec80fc89d951482d5c322f16ae9f2b4d5c (v2.9.1)

You no longer need the following (you can remove from your config):

 ```
 (add-to-list 'chatgpt-shell-system-prompts
               (cons "Mathematics"
                     "The user is a mathematician with very limited time.
You treat their time as precious. You do not repeat obvious things, including their query.
You are as concise as possible in responses.
You never apologize for confusions because it would waste their time.
All mathematical outputs must be in proper LaTeX format.
Use \\( and \\( for in-line equations, and \\[ and \\] for equation environments.
All mathematical delimiters must be on their own line.
All mathematical outputs are formally expressed in formal mathematical notation.
Don't give approximations or pronounciations for constants or variables.
You give formal definitions for all variables.
You keep your answers succinct and to-the-point.")
```
 or 

```
(add-hook 'chatgpt-shell-after-command-functions
          (lambda (command output success)
            (let ((file (buffer-file-name (buffer-base-buffer)))
                  ;; Silence org-element-at-point warnings.
                  (major-mode 'org-mode))
              (save-excursion
                (org-format-latex
                 (concat org-preview-latex-image-directory "org-ltximg")
                 (point-min) (point-max)
                 temporary-file-directory
                 'overlays nil 'forbuffer org-preview-latex-default-process)))))
```

The latex rendering can be removed via `chatgpt-shell-render-latex`.

Gonna close this issue as it seems to resolve the feature request. Please reopen if otherwise.

**Comment by xenodium** [comment]
ps. A couple of options improved the appearance for me:

```emacs-lisp
(setq org-format-latex-options (plist-put org-format-latex-options :scale 1.8))
(setq org-format-latex-options (plist-put org-format-latex-options :background "Transparent"))
```

<img width="774" alt="Screenshot 2024-12-18 at 11 31 46" src="https://github.com/user-attachments/assets/158bc43f-768a-482b-8900-a6ba14f05cfb" />

**Comment by kaibagley** [comment]
Amazing, thanks so much @xenodium!
